src/execution_engine.rs

```diff
@@ -432,59 +433,58 @@ impl Deref for ExecEngineInner {
     }
 }
 
-/// A wrapper around a function pointer which ensures the symbol being pointed
+/// A wrapper around a function pointer which ensures the function being pointed
 /// to doesn't accidentally outlive its execution engine.
 #[derive(Clone)]
-pub struct Symbol<F> {
-    _execution_engine: ExecEngineInner,
+pub struct JitFunction<'engine, F> {
+    _execution_engine: PhantomData<&'engine ExecutionEngine>,
     inner: F,
 }
 
-impl<F: UnsafeFunctionPointer> Deref for Symbol<F> {
-    type Target = F;
-
-    fn deref(&self) -> &Self::Target {
-        &self.inner
-    }
-}
-
-impl<F> Debug for Symbol<F> {
+impl<'engine, F> Debug for JitFunction<'engine, F> {
     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
-        f.debug_tuple("Symbol")
+        f.debug_tuple("JitFunction")
             .field(&"<unnamed>")
             .finish()
     }
 }
 
 /// Marker trait representing an unsafe function pointer (`unsafe extern "C" fn(A, B, ...) -> Output`).
-pub trait UnsafeFunctionPointer: private::Sealed + Copy {}
+pub trait UnsafeFunctionPointer: private::SealedUnsafeFunctionPointer {}
 
 mod private {
     /// A sealed trait which ensures nobody outside this crate can implement
     /// `UnsafeFunctionPointer`.
     ///
     /// See https://rust-lang-nursery.github.io/api-guidelines/future-proofing.html
-    pub trait Sealed {}
+    pub trait SealedUnsafeFunctionPointer: Copy {}
 }
 
+impl<F: private::SealedUnsafeFunctionPointer> UnsafeFunctionPointer for F {}
```

Comment:
Yeah, both are needed.

First clause on the trait simply requires that anything implementing `UnsafeFunctionPointer` has to also implement `SealedUnsafeFunctionPointer` (this way we guarantee that nobody outside the crate can implement the trait).

Second one defines blanket implementation for *all* `SealedUnsafeFunctionPointer` (in theory, and before this change, you can have different `impl`s on different types implementing `SealedUnsafeFunctionPointer`, but it's not very useful for marker-only type, so this allows to avoid duplication).